### PR TITLE
Feature/dtaller/add hip packages1

### PR DESCRIPTION
**Pull request by dtaller**
* Add Hip Support for umpire, chai, and raja
* Create HipPackage base class to support all hip builds and unify hip-specific variants in one place, mimicking the paradigm used for CudaPackage

Currently, the +hip variant works for camp, umpire, and raja

There is one build issue for chai+hip. We get complaints that the include path "/opt/rocm/hip/hcc/include" does not exist on the corona machine. This path is being incorrectly added to the include paths. However, I have determined that this is an issue with BLT rather than spack. In particular, see the offending lines:

blt/cmake/thirdparty/SetupHIP.cmake:25:set(HIP_RUNTIME_INCLUDE_DIRS "${HIP_ROOT_DIR}/include;${HIP_ROOT_DIR}/hcc/include" CACHE STRING "")
blt/host-configs/other/hcc.cmake:35:set(ROCM_INCLUDE_PATH "${ROCM_ROOT_DIR}/hcc/include;${ROCM_ROOT_DIR}/include"  CACHE PATH "")
 
A separate pull request will be submitted to the blt team shortly to deal with that issue

### CODE REVIEW

**Review by trws — COMMENTED** [review]
I like this overall. Assuming the TODOs get resolved LGTM.

**Comment by dtaller** [comment]
> I like this overall. Assuming the TODOs get resolved LGTM.

Thanks! Regarding the TODOs (issue of which versions of hip are allowed on different architectures etc.), you are right that I should follow up on that. I just sent a message to our contact on the hip team, and I am sure that he will respond with some documentation or clarification or contribution.

File: lib/spack/spack/build_systems/hip.py
Comment (robinson96) [inline lib/spack/spack/build_systems/hip.py]: has -> hsa

File: var/spack/repos/builtin/packages/chai/package.py
Comment (davidbeckingsale) [inline var/spack/repos/builtin/packages/chai/package.py]: I think you should use either `HIP_HIPCC_FLAGS` or `HIP_CLANG_FLAGS`. HCC is deprecated.

File: lib/spack/spack/build_systems/hip.py
Comment (dtaller) [reply lib/spack/spack/build_systems/hip.py]: Fixed

File: var/spack/repos/builtin/packages/chai/package.py
Comment (dtaller) [reply var/spack/repos/builtin/packages/chai/package.py]: I fixed it here and in camp, umpire, raja. The option HIP_HIPCC_FLAGS appears in a few other packages that I did not create, but I won't touch that during this pull request

**Comment by davidbeckingsale** [comment]
This looks good to me thanks @dtaller !

**Review by becker33 — APPROVED** [review]

**Review by srekolam — COMMENTED** [review]
after taking this change , we see failure in when we do spack install -v llvm-amdgpu@3.9.0, here is the log . When i revert back to older commit excluding this , i am able to build succesfully. I am looking into it now.
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:7:1: error: expected identifier or '(' before '.' token
 .section .text
 ^
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:16:5: error: invalid preprocessing directive #Save
   # Save scratch registers.
     ^~~~
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:44:5: error: invalid preprocessing directive #Align
   # Align stack frame.
     ^~~~~
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:45:14: error: stray '#' in program
   push %rbx  # non-scratch
              ^
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:48:19: error: stray '#' in program
   mov %rsp, %rbx  # save current rsp
                   ^
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:50:17: error: stray '#' in program
   shr $4, %rsp  # clear 4 lsb, align to 16
                 ^
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:55:5: error: invalid preprocessing directive #Unalign
   # Unalign stack frame back.
     ^~~~~~~
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:56:19: error: stray '#' in program
   mov %rbx, %rsp  # restore the original rsp
                   ^
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:60:5: error: invalid preprocessing directive #Restore
   # Restore scratch registers.
     ^~~~~~~
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:96:5: error: invalid preprocessing directive #Save
   # Save scratch registers.
     ^~~~
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:124:5: error: invalid preprocessing directive #Align
   # Align stack frame.
     ^~~~~
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:125:14: error: stray '#' in program
   push %rbx  # non-scratch
              ^
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:128:19: error: stray '#' in program
   mov %rsp, %rbx  # save current rsp
                   ^
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:130:17: error: stray '#' in program
   shr $4, %rsp  # clear 4 lsb, align to 16
                 ^
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:135:5: error: invalid preprocessing directive #Unalign
   # Unalign stack frame back.
     ^~~~~~~
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:136:19: error: stray '#' in program
   mov %rbx, %rsp  # restore the original rsp
                   ^
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:140:5: error: invalid preprocessing directive #Restore
   # Restore scratch registers.
     ^~~~~~~
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:212:41: error: stray '@' in program
   movq _ZN14__interception11real_setjmpE@GOTPCREL(%rip), %rdx
                                         ^
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:250:42: error: stray '@' in program
   movq _ZN14__interception12real__setjmpE@GOTPCREL(%rip), %rdx
                                          ^
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:308:44: error: stray '@' in program
   movq _ZN14__interception14real_sigsetjmpE@GOTPCREL(%rip), %rdx
                                            ^
/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-src/compiler-rt/lib/tsan/rtl/tsan_rtl_amd64.S:358:46: error: stray '@' in program
   movq _ZN14__interception16real___sigsetjmpE@GOTPCREL(%rip), %rdx
                                              ^
make[2]: *** [projects/compiler-rt/lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:475: projects/compiler-rt/lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_rtl_amd64.S.o] Error 1
make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-build-flacaut'
make[1]: *** [CMakeFiles/Makefile2:36691: projects/compiler-rt/lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/all] Error 2
make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-llvm-amdgpu-3.9.0-flacautny2vgojpuqcigxpyyl4m65ivb/spack-build-flacaut'
make: *** [Makefile:174: all] Error 2
==> Error: ProcessError: Command exited with status 2:
    'make'

**Comment by dtaller** [comment]
Hi @srekolam . I am sorry about that.

My suggestion would be to check the paths in var/spack/repos/builtin/packages/hip/package.py  . I made some minor changes. I was only able to check these changes in cases where the package was external, which may not be the case for you. I would check the function get_rocm_prefix_info in particular, and the places where it is called.

**Comment by balos1** [comment]
@srekolam I was also running into these same errors, but I found that the issue is actually CMake 3.19.0 (see links) not this PR. It is likely that when you reverted to an earlier commit, you reverted to a point before cmake@3.19.0 was added to spack (which [happened one day after this PR][1]) and that is why it seemed like this PR was the problem. 

I was able to get past the errors by using cmake@3.18.4 instead.

For reference:
https://gitlab.kitware.com/cmake/cmake/-/issues/21469
https://bugs.gentoo.org/756034


[1]: https://github.com/spack/spack/pull/19996

**Comment by srekolam** [comment]
@balos1  thanks for the update. i verified that with this commit, i am able to build llvm-amdgpu @3.9.0 and hip@3.9.0 but the rocprim@3.9.0 or other recipes were not building for us due to some of changes in the hip package..  At this point we were trying to find 2 things - llvm-amdgpu package failure on the latest ,  rocprim@3.9.0 and other packages changes with this PR
For the second , it is likely inside the setup_dependent_build_environment() in the hip package, if we use below it appears to resolve the build issue for rocmprim, rocsparse etc..
if self.spec.external:
            env.set('DEVICE_LIB_PATH', rocm_prefixes['rocm-device-libs'].lib)
            env.set('HIP_PATH', rocm_prefixes['rocm-path'])
        else:
            env.set('DEVICE_LIB_PATH', rocm_prefixes['rocm-device-libs'].amdgcn.bitcode)
            env.set('HIP_PATH', self.prefix)

**Comment by dtaller** [comment]
I'm glad that you got to the bottom of this, @srekolam . Do you want to make a PR with those changes? I'm sorry for the trouble

File: var/spack/repos/builtin/packages/hip/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/hip/package.py:124]: This is probably incorrect @dtaller, can you please test with all packages installed by spack instead of by the system?

File: var/spack/repos/builtin/packages/hip/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/hip/package.py:131]: This is incorrect since arch can be tuple

**Comment by haampie** [comment]
I didn't have the time to review this PR, and it seems it wasn't thoroughly tested before it got merged :(.

@dtaller, are you OK with renaming `+hip` to `+rocm` and maybe HipPackage to ROCmPackage? Those names were already used by `spfft`, `sirius` and `spla`.

Also there should be support for multiple amd gpu architectures, like in `spfft`, `sirius` etc.

File: var/spack/repos/builtin/packages/hip/package.py
Comment (haampie) [reply var/spack/repos/builtin/packages/hip/package.py:131]: Ok, I see that multi=True part was not copied over from existing packages...

File: var/spack/repos/builtin/packages/chai/package.py
Comment (haampie) [reply var/spack/repos/builtin/packages/chai/package.py]: There isn't such a thing as `HIP_HIPCC_FLAGS` or `HIP_CLANG_FLAGS` in `hipcc`. Edit: ah, I see, this is just flags for cmake.

**Comment by dtaller** [comment]
Hi @haampie . Thank you for your comments. Yes, I am ok with these changes, and I am OK with changing +hip to +rocm and HipPackage to ROCmPackage , most likely after your other pull request goes in (see https://github.com/spack/spack/pull/20095 ).

I would advise adding another reviewer to your pull request because I am very new user/contributor to spack and I certainly don't have  privileges to merge it in once approved in any case

**Pull request merged**
